File: php-sdk/src/CleansterClient.php
<?php

declare(strict_types=1);

namespace Cleanster;

use Cleanster\Api\BlacklistApi;
use Cleanster\Api\BookingsApi;
use Cleanster\Api\ChecklistsApi;
use Cleanster\Api\OtherApi;
use Cleanster\Api\PaymentMethodsApi;
use Cleanster\Api\PropertiesApi;
use Cleanster\Api\UsersApi;
use Cleanster\Api\WebhooksApi;

/**
 * Main entry point for the Cleanster Partner API SDK.
 *
 * Create a client using the sandbox() or production() factory methods.
 * After creating a user, call setAccessToken() with the user bearer token
 * so it is automatically included on every subsequent request.
 *
 * ```php
 * $client = CleansterClient::sandbox($_ENV['CLEANSTER_API_KEY']);
 *
 * $userResp = $client->users()->createUser('jane@example.com', 'Jane', 'Smith');
 * $token    = $client->users()->fetchAccessToken($userResp->data->id);
 * $client->setAccessToken($token->data->token);
 *
 * $bookings = $client->bookings()->getBookings();
 * ```
 */
final class CleansterClient
{
    private readonly HttpClient      $http;
    private readonly BookingsApi     $bookings;
    private readonly UsersApi        $users;
    private readonly PropertiesApi   $properties;
    private readonly ChecklistsApi   $checklists;
    private readonly OtherApi        $other;
    private readonly BlacklistApi    $blacklist;
    private readonly PaymentMethodsApi $paymentMethods;
    private readonly WebhooksApi     $webhooks;

    /**
     * @param Config          $config     SDK configuration.
     * @param HttpClient|null $httpClient Optional custom HTTP client — used in tests.
     */
    public function __construct(Config $config, ?HttpClient $httpClient = null)
    {
        $this->http           = $httpClient ?? new HttpClient($config);
        $this->bookings       = new BookingsApi($this->http);
        $this->users          = new UsersApi($this->http);
        $this->properties     = new PropertiesApi($this->http);
        $this->checklists     = new ChecklistsApi($this->http);
        $this->other          = new OtherApi($this->http);
        $this->blacklist      = new BlacklistApi($this->http);
        $this->paymentMethods = new PaymentMethodsApi($this->http);
        $this->webhooks       = new WebhooksApi($this->http);
    }

    /** Create a client configured for the sandbox environment (no real charges). */
    public static function sandbox(string $accessKey): self
    {
        return new self(Config::sandbox($accessKey));
    }

    /** Create a client configured for the production environment (live traffic). */
    public static function production(string $accessKey): self
    {
        return new self(Config::production($accessKey));
    }

    /**
     * Set the user bearer token.
     *
     * Call this after fetching a user token with users()->fetchAccessToken().
     * The token is sent as the "token" HTTP header on every subsequent request.
     *
     * @param string $token User bearer token from fetchAccessToken().
     */
    public function setAccessToken(string $token): void
    {
        $this->http->setToken($token);
    }

    /** Return the currently active bearer token, or '' if not set. */
    public function getAccessToken(): string
    {
        return $this->http->getToken();
    }

    /** Access booking management methods. */
    public function bookings(): BookingsApi
    {
        return $this->bookings;
    }

    /** Access user account and authentication methods. */
    public function users(): UsersApi
    {
        return $this->users;
    }

    /** Access property management methods. */
    public function properties(): PropertiesApi
    {
        return $this->properties;
    }

    /** Access checklist management methods. */
    public function checklists(): ChecklistsApi
    {
        return $this->checklists;
    }

    /** Access utility/reference data methods (services, plans, cost estimates, etc.). */
    public function other(): OtherApi
    {
        return $this->other;
    }

    /** Access cleaner blacklist management methods. */
    public function blacklist(): BlacklistApi
    {
        return $this->blacklist;
    }

    /** Access payment method management methods. */
    public function paymentMethods(): PaymentMethodsApi
    {
        return $this->paymentMethods;
    }

    /** Access webhook management methods. */
    public function webhooks(): WebhooksApi
    {
        return $this->webhooks;
    }
}

File: csharp-sdk/src/Cleanster/CleansterClient.cs
using Cleanster.Api;

namespace Cleanster;

/// <summary>
/// Main entry point for the Cleanster Partner API SDK.
///
/// Create via <see cref="Sandbox"/> or <see cref="Production"/> factory methods.
/// After creating a user, call <see cref="SetAccessToken"/> with the user bearer token
/// so every subsequent request includes it automatically.
///
/// <code>
/// var client = CleansterClient.Sandbox(Environment.GetEnvironmentVariable("CLEANSTER_API_KEY")!);
///
/// var userResp  = await client.Users.CreateUserAsync("jane@example.com", "Jane", "Smith");
/// var tokenResp = await client.Users.FetchAccessTokenAsync(userResp.Data.Id);
/// client.SetAccessToken(tokenResp.Data.Token!);
///
/// var bookings = await client.Bookings.GetBookingsAsync();
/// </code>
/// </summary>
public sealed class CleansterClient : IDisposable
{
    private readonly ICleansterHttpClient _http;

    /// <summary>Booking lifecycle and management.</summary>
    public BookingsApi        Bookings       { get; }

    /// <summary>User account creation and bearer-token management.</summary>
    public UsersApi           Users          { get; }

    /// <summary>Property (location) management.</summary>
    public PropertiesApi      Properties     { get; }

    /// <summary>Cleaning task list management.</summary>
    public ChecklistsApi      Checklists     { get; }

    /// <summary>Reference data and utility endpoints.</summary>
    public OtherApi           Other          { get; }

    /// <summary>Cleaner blacklist management.</summary>
    public BlacklistApi       Blacklist      { get; }

    /// <summary>Stripe and PayPal payment method management.</summary>
    public PaymentMethodsApi  PaymentMethods { get; }

    /// <summary>Webhook endpoint management.</summary>
    public WebhooksApi        Webhooks       { get; }

    /// <summary>
    /// Create a client using an explicit config and optional custom HTTP transport.
    /// Pass a custom <paramref name="httpClient"/> for unit testing.
    /// </summary>
    public CleansterClient(CleansterConfig config, ICleansterHttpClient? httpClient = null)
    {
        _http          = httpClient ?? new CleansterHttpClient(config);
        Bookings       = new BookingsApi(_http);
        Users          = new UsersApi(_http);
        Properties     = new PropertiesApi(_http);
        Checklists     = new ChecklistsApi(_http);
        Other          = new OtherApi(_http);
        Blacklist      = new BlacklistApi(_http);
        PaymentMethods = new PaymentMethodsApi(_http);
        Webhooks       = new WebhooksApi(_http);
    }

    /// <summary>Create a client configured for the sandbox environment (no real charges).</summary>
    public static CleansterClient Sandbox(string accessKey)
        => new(CleansterConfig.Sandbox(accessKey));

    /// <summary>Create a client configured for the production environment (live traffic).</summary>
    public static CleansterClient Production(string accessKey)
        => new(CleansterConfig.Production(accessKey));

    /// <summary>
    /// Set the user bearer token for all subsequent requests.
    ///
    /// Call this after <see cref="UsersApi.FetchAccessTokenAsync"/> returns a token.
    /// The token is long-lived — store it and reuse it across requests.
    /// </summary>
    public void SetAccessToken(string token) => _http.SetToken(token);

    /// <summary>Return the currently active bearer token, or <see cref="string.Empty"/> if not set.</summary>
    public string GetAccessToken() => _http.GetToken();

    /// <inheritdoc/>
    public void Dispose() => _http.Dispose();
}
